File: app/src/main/java/me/zhanghai/android/files/provider/document/DocumentFileSystem.java
/*
 * Copyright (c) 2019 Hai Zhang <dreaming.in.code.zh@gmail.com>
 * All Rights Reserved.
 */

package me.zhanghai.android.files.provider.document;

import android.net.Uri;
import android.os.Parcel;
import android.os.Parcelable;

import java.io.IOException;
import java.util.Collections;
import java.util.Objects;
import java.util.Set;

import androidx.annotation.NonNull;
import java8.nio.file.FileStore;
import java8.nio.file.FileSystem;
import java8.nio.file.Path;
import java8.nio.file.PathMatcher;
import java8.nio.file.WatchService;
import java8.nio.file.attribute.UserPrincipalLookupService;
import java8.nio.file.spi.FileSystemProvider;
import me.zhanghai.android.files.provider.common.ByteString;
import me.zhanghai.android.files.provider.common.ByteStringBuilder;
import me.zhanghai.android.files.provider.common.ByteStringListPathFactory;

class DocumentFileSystem extends FileSystem implements ByteStringListPathFactory, Parcelable {

    static final byte SEPARATOR = '/';

    private static final ByteString SEPARATOR_BYTE_STRING = ByteString.ofByte(SEPARATOR);
    private static final String SEPARATOR_STRING = Character.toString((char) SEPARATOR);

    @NonNull
    private final DocumentFileSystemProvider mProvider;

    @NonNull
    private final Uri mTreeUri;

    @NonNull
    private final DocumentPath mRootDirectory;

    @NonNull
    private final Object mLock = new Object();

    private boolean mOpen = true;

    DocumentFileSystem(@NonNull DocumentFileSystemProvider provider, @NonNull Uri treeUri) {
        mProvider = provider;
        mTreeUri = treeUri;

        mRootDirectory = new DocumentPath(this, SEPARATOR_BYTE_STRING);
        if (!mRootDirectory.isAbsolute()) {
            throw new AssertionError("Root directory must be absolute");
        }
        if (mRootDirectory.getNameCount() != 0) {
            throw new AssertionError("Root directory must contain no names");
        }
    }

    @NonNull
    DocumentPath getRootDirectory() {
        return mRootDirectory;
    }

    @NonNull
    DocumentPath getDefaultDirectory() {
        return mRootDirectory;
    }

    @NonNull
    Uri getTreeUri() {
        return mTreeUri;
    }

    @NonNull
    @Override
    public FileSystemProvider provider() {
        return mProvider;
    }

    @Override
    public void close() {
        synchronized (mLock) {
            if (!mOpen) {
                return;
            }
            mProvider.removeFileSystem(this);
            mOpen = false;
        }
    }

    @Override
    public boolean isOpen() {
        synchronized (mLock) {
            return mOpen;
        }
    }

    @Override
    public boolean isReadOnly() {
        return false;
    }

    @NonNull
    @Override
    public String getSeparator() {
        return SEPARATOR_STRING;
    }

    @NonNull
    @Override
    public Iterable<Path> getRootDirectories() {
        return Collections.singletonList(mRootDirectory);
    }

    @NonNull
    @Override
    public Iterable<FileStore> getFileStores() {
        // TODO
        throw new UnsupportedOperationException();
    }

    @NonNull
    @Override
    public Set<String> supportedFileAttributeViews() {
        return DocumentFileAttributeView.SUPPORTED_NAMES;
    }

    @NonNull
    @Override
    public DocumentPath getPath(@NonNull String first, @NonNull String... more) {
        Objects.requireNonNull(first);
        Objects.requireNonNull(more);
        ByteStringBuilder pathBuilder = new ByteStringBuilder(ByteString.fromString(first));
        for (String name : more) {
            Objects.requireNonNull(name);
            pathBuilder
                    .append(SEPARATOR)
                    .append(ByteString.fromString(name));
        }
        ByteString path = pathBuilder.toByteString();
        return new DocumentPath(this, path);
    }

    @NonNull
    @Override
    public DocumentPath getPath(@NonNull ByteString first, @NonNull ByteString... more) {
        Objects.requireNonNull(first);
        Objects.requireNonNull(more);
        ByteStringBuilder pathBuilder = new ByteStringBuilder(first);
        for (ByteString name : more) {
            Objects.requireNonNull(name);
            pathBuilder
                    .append(SEPARATOR)
                    .append(name);
        }
        ByteString path = pathBuilder.toByteString();
        return new DocumentPath(this, path);
    }

    @NonNull
    @Override
    public PathMatcher getPathMatcher(String syntaxAndPattern) {
        Objects.requireNonNull(syntaxAndPattern);
        throw new UnsupportedOperationException();
    }

    @NonNull
    @Override
    public UserPrincipalLookupService getUserPrincipalLookupService() {
        throw new UnsupportedOperationException();
    }

    @NonNull
    @Override
    public WatchService newWatchService() throws IOException {
        // TODO
        throw new UnsupportedOperationException();
    }

    @Override
    public boolean equals(Object object) {
        if (this == object) {
            return true;
        }
        if (object == null || getClass() != object.getClass()) {
            return false;
        }
        DocumentFileSystem that = (DocumentFileSystem) object;
        return Objects.equals(mTreeUri, that.mTreeUri);
    }

    @Override
    public int hashCode() {
        return Objects.hash(mTreeUri);
    }


    public static final Creator<DocumentFileSystem> CREATOR = new Creator<DocumentFileSystem>() {
        @Override
        public DocumentFileSystem createFromParcel(Parcel source) {
            Uri treeUri = source.readParcelable(Path.class.getClassLoader());
            return DocumentFileSystemProvider.getOrNewFileSystem(treeUri);
        }
        @Override
        public DocumentFileSystem[] newArray(int size) {
            return new DocumentFileSystem[size];
        }
    };

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeParcelable(mTreeUri, flags);
    }
}

File: app/src/main/java/me/zhanghai/android/files/provider/document/DocumentFileSystem.kt
/*
 * Copyright (c) 2019 Hai Zhang <dreaming.in.code.zh@gmail.com>
 * All Rights Reserved.
 */

package me.zhanghai.android.files.provider.document

import android.net.Uri
import android.os.Parcel
import android.os.Parcelable
import java8.nio.file.FileStore
import java8.nio.file.FileSystem
import java8.nio.file.Path
import java8.nio.file.PathMatcher
import java8.nio.file.WatchService
import java8.nio.file.attribute.UserPrincipalLookupService
import java8.nio.file.spi.FileSystemProvider
import me.zhanghai.android.files.provider.common.ByteString
import me.zhanghai.android.files.provider.common.ByteStringBuilder
import me.zhanghai.android.files.provider.common.ByteStringListPathCreator
import me.zhanghai.android.files.provider.common.toByteString
import me.zhanghai.android.files.util.StableUriParceler
import java.io.IOException

internal class DocumentFileSystem(
    private val provider: DocumentFileSystemProvider,
    val treeUri: Uri
) : FileSystem(), ByteStringListPathCreator, Parcelable {
    val rootDirectory = DocumentPath(this, SEPARATOR_BYTE_STRING)

    init {
        if (!rootDirectory.isAbsolute) {
            throw AssertionError("Root directory must be absolute")
        }
        if (rootDirectory.nameCount != 0) {
            throw AssertionError("Root directory must contain no names")
        }
    }

    private val lock = Any()

    private var isOpen = true

    val defaultDirectory: DocumentPath
        get() = rootDirectory

    override fun provider(): FileSystemProvider = provider

    override fun close() {
        synchronized(lock) {
            if (!isOpen) {
                return
            }
            provider.removeFileSystem(this)
            isOpen = false
        }
    }

    override fun isOpen(): Boolean = synchronized(lock) { isOpen }

    override fun isReadOnly(): Boolean = false

    override fun getSeparator(): String = SEPARATOR_STRING

    override fun getRootDirectories(): Iterable<Path> = listOf(rootDirectory)

    override fun getFileStores(): Iterable<FileStore> {
        // TODO
        throw UnsupportedOperationException()
    }

    override fun supportedFileAttributeViews(): Set<String> =
        DocumentFileAttributeView.SUPPORTED_NAMES

    override fun getPath(first: String, vararg more: String): DocumentPath {
        val path = ByteStringBuilder(first.toByteString())
            .apply { more.forEach { append(SEPARATOR).append(it.toByteString()) } }
            .toByteString()
        return DocumentPath(this, path)
    }

    override fun getPath(first: ByteString, vararg more: ByteString): DocumentPath {
        val path = ByteStringBuilder(first)
            .apply { more.forEach { append(SEPARATOR).append(it) } }
            .toByteString()
        return DocumentPath(this, path)
    }

    override fun getPathMatcher(syntaxAndPattern: String): PathMatcher {
        throw UnsupportedOperationException()
    }

    override fun getUserPrincipalLookupService(): UserPrincipalLookupService {
        throw UnsupportedOperationException()
    }

    @Throws(IOException::class)
    override fun newWatchService(): WatchService {
        // TODO
        throw UnsupportedOperationException()
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) {
            return true
        }
        if (javaClass != other?.javaClass) {
            return false
        }
        other as DocumentFileSystem
        return treeUri == other.treeUri
    }

    override fun hashCode(): Int = treeUri.hashCode()

    override fun describeContents(): Int = 0

    override fun writeToParcel(dest: Parcel, flags: Int) {
        //dest.writeParcelable(treeUri, flags)
        with(StableUriParceler) { treeUri.write(dest, flags) }
    }

    companion object {
        const val SEPARATOR = '/'.code.toByte()
        private val SEPARATOR_BYTE_STRING = SEPARATOR.toByteString()
        private const val SEPARATOR_STRING = SEPARATOR.toInt().toChar().toString()

        @JvmField
        val CREATOR = object : Parcelable.Creator<DocumentFileSystem> {
            override fun createFromParcel(source: Parcel): DocumentFileSystem {
                //val treeUri = source.readParcelable<Uri>()!!
                val treeUri = StableUriParceler.create(source)!!
                return DocumentFileSystemProvider.getOrNewFileSystem(treeUri)
            }

            override fun newArray(size: Int): Array<DocumentFileSystem?> = arrayOfNulls(size)
        }
    }
}
